### PR TITLE
Partial JSON types to enable strict null checks

### PR DESCRIPTION
**Pull request by vidartf**
Proof of concept for having partial JSON types, allowing strict null checks to be enabled.

## References

Proof of concept, alternative resolution for https://github.com/phosphorjs/phosphor/pull/303. I.e. if this concept is acceptable for inclusion in phosphor it should be upstreamed.

## Code changes

Adds types `PartialJSONObject` etc that allow undefined values / missing keys on the JSON objects.

Also enables `strictNullChecks` on coreutils and services. There is currently one compile error, which was an error revealed by the strict checks (i.e. not introduced by this PR).

## User-facing changes

None, this is purely about type safety, but it might fix some bugs.

## Backwards-incompatible changes

Consumers would possibly have to change some imports to be compatible with APIs, but this should be considered a "bug fix", as it is simply clarifying the currently in-use types.

### CODE REVIEW

**Comment by blink1073** [comment]
Oooh, strict nulls!!!

**Comment by vidartf** [comment]
CC @afshin as I did some fixes in code he wrote, and I'd like some checks if I did it right 🤞

File: packages/coreutils/src/partialjson.ts [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+   * A shared frozen empty JSONArray
+   */
+  export const emptyArray = Object.freeze([]) as ReadonlyPartialJSONArray;
+
+  /**
+   * Test whether a JSON value is a primitive.
+   *
+   * @param value - The JSON value of interest.
+   *
+   * @returns `true` if the value is a primitive,`false` otherwise.
```
Comment (blink1073) [inline packages/coreutils/src/partialjson.ts]: missing a space after the comma

**Comment by blink1073** [comment]
The approach and implementation look good to me.  I also think we should have the `PartialJSON` logic in Phosphor itself.

**Comment by vidartf** [comment]
> implementation look good to me.

Well, it is your code... 

image

**Comment by ian-r-rose** [comment]
gotothere

**Comment by blink1073** [comment]
Gosh darn it!

**Comment by jasongrout** [comment]
What is the current status of this PR?

**Comment by vidartf** [comment]
It looked like the Phosphor PR was moving, so this was put on ice. Now it seems to have stalled, so maybe we should start this up again?

**Comment by blink1073** [comment]
Closing in favor of #7607

**Pull request closed without merging**